assets/js/base/hooks/use-shipping-rates.js

```diff
@@ -1,40 +1,63 @@
 /**
  * External dependencies
  */
-import { useSelect } from '@wordpress/data';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useReducer, useEffect } from '@wordpress/element';
+import isShallowEqual from '@wordpress/is-shallow-equal';
+import { useDebounce } from 'use-debounce';
 import { CART_STORE_KEY as storeKey } from '@woocommerce/block-data';
 /**
  * Internal dependencies
  */
 import { useStoreCart } from './use-store-cart';
-
+import { pluckAddress } from '../utils';
 /**
- * This is a custom hook that is wired up to the `wc/store/collections` data
- * store for the `wc/store/cart/shipping-rates` route. Given a query object, this
- * will ensure a component is kept up to date with the shipping rates matching that
- * query in the store state.
+ * This is a custom hook that is wired up to the `wc/store/cart/shipping-rates` route.
+ * Given a a set of default fields keys, this will handle shipping form state and load
+ * new rates when certain fields change.
+ *
+ * @param addressFields an object containing default fields keys.
  *
  * @return {Object} This hook will return an object with three properties:
- *                  - shippingRates        An array of shipping rate objects.
- *                  - shippingRatesLoading A boolean indicating whether the shipping
- *                                         rates are still loading or not.
- *                  - updateShipping       An action dispatcher to update the shipping address.
+ *                 - {Boolean} shippingRatesLoading A boolean indicating whether the shipping
+ *                   rates are still loading or not.
+ *                 - {Function} setShippingAddress  An function that optimistically
+ *                   update shipping address and dispatches async rate fetching.
+ *                 - {Object} shippingAddress       An object containing shipping address.
  */
-export const useShippingRates = () => {
+export const useShippingRates = ( addressFields ) => {
 	const { shippingRates } = useStoreCart();
-	const results = useSelect( ( select, { dispatch } ) => {
-		const store = select( storeKey );
-		const shippingRatesLoading = store.areShippingRatesLoading();
-		const { updateShippingAddress } = dispatch( storeKey );
-
-		return {
-			shippingRatesLoading,
-			updateShippingAddress,
-		};
-	}, [] );
+	const derivedAddress = shippingRates[ 0 ]?.destination;
+	const initialAddress = { ...addressFields, ...derivedAddress };
+	const shippingAddressReducer = ( state, address ) => ( {
+		...state,
+		...address,
+	} );
+	const [ shippingAddress, setShippingAddress ] = useReducer(
```

Comment:
One issue that is not blocking but didn’t understand yet is that `useReducer` initial value doesn’t work like `useState`, it seems `shippingAddress` is returning empty values, even if `initialAddress` has values in it, this means on page load, no address will be in the checkout shipping fields, this is not blocking but we should fix it.